### PR TITLE
Track `ast::Param`s that output types are lifetime bound to in `ast::Method`

### PR DESCRIPTION
**Pull request by QnnOkabayashi**
This PR introduces `Method::output_lifetime_dependent_params`, providing the ability for `ast::Method`s to track which `ast::Param`s their output types are lifetime bound to:
```rust
impl Method {
    pub fn output_lifetime_dependent_params(&self) -> Vec<&Param> { ... }
}
```

For GC languages like JS, this will allow us to hold references to these params to prevent the GC from dropping them later. For non-GC languages like C++, this lets us generate documentation that specifies which parameters the return type is lifetime-bound to.
___
I'm opening this as a draft PR because I still have some unresolved questions that I left in the form of comments. I also think there's room for improvement on naming the `Method::output_lifetime_dependent_params` method. Some initial ideas:
* `Method::return_bound_params`
* `Method::lifetime_bound_params`
* `Method::params_held_by_output`
* `Method::params_lifetime_bound_by_return_type`
* `Method::params_that_must_live_as_long_as_return_type`

After this gets merged, the next steps will be to use this information in language-specific backends. I'm planning to start with the JS backend, and then tackle the C++ one later. We can also consider supporting lifetime elision at a later point, but that's disjoint from the language support.

### CODE REVIEW

File: core/src/ast/types.rs [hunk 1]
```diff
@@ -641,6 +643,74 @@ impl TypeName {
         self.check_option(errors);
     }
 
+    /// Checks the validity of return types.
+    ///
+    /// This is equivalent to `TypeName::check_validity`, but it also ensures
+    /// that the type doesn't elide any lifetimes.
+    ///
+    /// Once we decide to support lifetime elision in return types, this function
```
Comment (Manishearth) [inline core/src/ast/types.rs:641]: nit: worth filing a followup for this and linking to it

File: core/src/ast/methods.rs [hunk 2]
```diff
@@ -102,6 +104,52 @@ impl Method {
         }
     }
 
+    /// Returns the parameters that the output type's lifetime is bound to.
+    ///
+    /// WARNING: This may produce incorrect results if the validity isn't checked
+    /// first, since the result type may contain elided lifetimes that we depend
```
Comment (Manishearth) [inline core/src/ast/methods.rs]: nit: worth mentioning _which_ validity

**Review by Manishearth — CHANGES_REQUESTED** [review]
looks pretty good!


some small comments

File: core/src/ast/types.rs (hunk 1 above)
Comment (QnnOkabayashi) [reply core/src/ast/types.rs:641]: Done: #160

Comment (Manishearth) [reply core/src/ast/types.rs:641]: (in the code as well, you can also just mention the issue number)

File: core/src/ast/methods.rs [hunk 3]
```diff
@@ -302,4 +303,54 @@ mod tests {
             vec![]
         ));
     }
+
+    #[test]
+    fn static_params_held_by_return_type() {
```
[the remark this answers is not in the document]
Comment (QnnOkabayashi) [reply core/src/ast/methods.rs:336]: Makes sense, I made it use assertions now

**Review by Manishearth — APPROVED** [review]

**Pull request merged**